[19759] Port category filter to OP 4.1+

floriank opened the pull request:
This will port the work @NobodysNightmare put into backporting a missing feature to the `release/4.0` branch into `release/4.1` while accommodating the new file structure for the frontend. 

Tests have been added where necessary, I am not yet sure if we should add some integration tests for filtering the work package list.

The feature to filter by category originally got lost while developing version 4.0 and is now back through the efforts in #2908.

This should meet the requirements of https://community.openproject.org/work_packages/19759.

public WP: https://community.openproject.org/work_packages/19809

NobodysNightmare committed:
Backporting APIv3 categories to OP 4.0

Backport was neccessary to implement missing categories filter without adding more technical debt.
These changes will be merged into 4.1 and following branches and vanish in that merge (merge-strategy "ours")

YOU SHOULD NOT SEE ANY CONTENTS FROM THIS COMMIT ANYWHERE AFTER 4.0.x!

This commit was built from the commits of the GitHub-PR mentioned below, as well as some code dependencies for that.

The GitHub-PR was https://github.com/opf/openproject/pull/2505

NobodysNightmare committed:
WIP wire up category filters in frontend

Commit:
parse categories from api response

finish up the groundwork laid out by @NobodysNightmare and actually parse the
response fromthe backported api

Commit:
add the category service from the backport

Commit:
backport the category filter from 4.0

This backports (forward ports) the frontend implementation of the category
filter to the new angular based frontend in 4.1

Commit:
add a path helper test

checks the newly added paths for the PathHelper

Commit:
add a test for the newly added category service

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://floriank-feature-19759-fix-category-filter.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-feature-19759-fix-category-filter.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc commented on frontend/tests/unit/tests/services/category-service-test.js:
```diff
@@ -0,0 +1,75 @@
+      $httpBackend.expectGET(apiPath);
+
+      var callback = sinon.spy(),
+          requestedCategories = CategoryService.getCategories(projectIdentifier).then(callback);
```
@floriank not critical, but what the Hound says.

floriank replied:
i noticed it locally using `jshint`\- i'll put in a styling commit after i have coped with some other issues.

NobodysNightmare replied:
@floriank The time has come ;-)

Azure7111 commented:
Front-end wise this looks good. There isn't even a migration necessary to keep using old category filters from `release/3.0`.

Somebody else will have to actually review the code though, sorry :/.

floriank commented:
Note: Public WP - https://community.openproject.org/work_packages/19809

NobodysNightmare commented:
Does not yet merge the correct 4.0 commits.

I will remove the WIP label as soon as it does.

NobodysNightmare committed:
Resolving forward-merge of backport PR

Most changes applied to 4.0.x shall not be applied to 4.1, the rest will be applied manually.

backport PR was: https://github.com/opf/openproject/pull/2908

NobodysNightmare commented on frontend/tests/unit/tests/helpers/path-helper-test.js:
```diff
@@ -0,0 +1,49 @@
+  }));
+
+  context('apiV3', function() {
+    var projectIdentifier = 'majora';
```
:+1:

Fine to see no specs for this helper until now ^^

Commit:
fix issues found by hound

[ci skip]
